src/types/arrow.rs

```diff
@@ -18,19 +18,30 @@ use crate::node::{
     CoreConstructible, DisconnectConstructible, JetConstructible, NoDisconnect,
     WitnessConstructible,
 };
-use crate::types::{Bound, Error, Final, Type};
+use crate::types::{Bound, Context, Error, Final, Type};
 use crate::{jet::Jet, Value};
 
 use super::variable::new_name;
 
 /// A container for an expression's source and target types, whether or not
 /// these types are complete.
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct Arrow {
     /// The source type
     pub source: Type,
     /// The target type
     pub target: Type,
+    /// Type inference context for both types.
+    pub inference_context: Context,
+}
+
+// Having `Clone` makes it easier to derive Clone on structures
+// that contain Arrow, even though it is potentially confusing
+// to use `.clone` to mean a shallow clone.
+impl Clone for Arrow {
+    fn clone(&self) -> Self {
+        self.shallow_clone()
+    }
```

Comment:
8eeab8f2e6dc61eba299e57b282ecf677a9bce0a: Would it make sense to introduce a `ShallowClone` trait with `Clone` as a subtrait? `ShallowClone` would be almost like a marker trait and it would provide the method `shallow_clone` which may call `clone` by default. If people see that both `ShallowClone` and `Clone` is implemented for a type, then they can infer that the type is cheaply cloneable

Reply:
Promoted to https://github.com/BlockstreamResearch/rust-simplicity/issues/232. I like this idea but it's separate from this PR really.

cc https://smallcultfollowing.com/babysteps/blog/2024/06/21/claim-auto-and-otherwise/